pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,332 @@
+			// if it is not active already (aescbc as first provider) then set it to active
+			// and remember to write this created configuration
+			b.Logger.Info("activating existing passive EncryptionConfiguration")
+			encryptionconfiguration.SetActive(ecSeed, true)
+			needToWriteConfig = true
+		}
+	}
+	ecSeedYamlBytes, err := encryptionconfiguration.ToYAML(ecSeed)
+	if err != nil {
+		return err
+	}
+	if needToWriteConfig {
+		b.Logger.Info("writing new/updated EncryptionConfiguration")
+		err = b.writeEncryptionConfiguration(ecSeedYamlBytes)
+		if err != nil {
+			return err
+		}
+	}
+
+	checksum := utils.ComputeSHA256Hex(ecSeedYamlBytes)
+	b.mutex.Lock()
+	b.CheckSums[common.EtcdEncryptionSecretName] = checksum
+	b.mutex.Unlock()
+	// enablement of etcd encryption feature done in helm chart of apiserver deployment
+	return nil
+}
+
+// RewriteShootSecretsIfEncryptionConfigurationChanged rewrites a shoot's secrets if the EncryptionConfiguration has changed
+func (b *Botanist) RewriteShootSecretsIfEncryptionConfigurationChanged() error {
+	b.Logger.Info("starting RewriteShootSecretsIfEncryptionConfigurationChanged")
+	// WARNING:
+	// No explicit checking of whether EncryptionConfiguration is contained in a backup.
+	// Be aware of the risk!
+	//
+	// TODO: Ensure this is also agreed upon by Gardener team
+
+	// TODO: contact Amshuman Rao Karaya
+
+	enableEtcdEncryption, err := utils.CheckVersionMeetsConstraint(b.Shoot.Info.Spec.Kubernetes.Version, ">= 1.13")
+	if err != nil {
+		return err
+	}
+	if !enableEtcdEncryption {
+		return nil
+	}
+	needToRewrite, err := b.needToRewriteShootSecrets()
+	if err != nil {
+		return err
+	}
+	if needToRewrite {
+		b.Logger.Info("rewriting shoot secrets")
+		err = b.rewriteShootSecrets()
+		if err != nil {
+			return err
+		}
+	} else {
+		b.Logger.Trace("EncryptionConfiguration unchanged, no rewriting necessary")
+	}
+	return nil
+}
+
+// readEncryptionConfigurationFromSeed reads the EncryptionConfiguration from the shoot namespace in the seed
+func (b *Botanist) readEncryptionConfigurationFromSeed() (exists bool, ec *apiserverconfigv1.EncryptionConfiguration, forcePlaintext bool, err error) {
+	client := b.Operation.K8sSeedClient
+	ecs, err := client.GetSecret(b.Operation.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil, false, nil
+		} else {
+			return false, nil, false, err
+		}
+	}
+	secretData, ok := ecs.Data[common.EtcdEncryptionSecretFileName]
+	if !ok {
+		return true, nil, false, fmt.Errorf("EncryptionConfiguration in seed cluster (%v) does not contain expected element: %v", common.EtcdEncryptionSecretName, common.EtcdEncryptionSecretFileName)
+	}
+	ec, err = encryptionconfiguration.CreateFromYAML(secretData)
+	if err != nil {
+		return true, nil, false, fmt.Errorf("EncryptionConfiguration in seed cluster (%v) is not consistent: %v", common.EtcdEncryptionSecretName, err)
```

Comment:
Not sure whether I understand the error message here - how is an error in "creating from yaml" related to "not consistent"?

Reply:
well, the UniversalDecoder().Decode fails when trying to decode the byte array ... since this is actually the configuration which I create myself, I can imagine this error only if somebody manually interfered with the configuration, which is why I thought that the configuration is no longer consistent ...

Reply:
I adapted the loading of the configuration as proposed by Rafael, hope this satisfies this comment as well ...